django_emarsys/event.py

```diff
@@ -0,0 +1,263 @@
+
+
+def get_parameter_for_event(event_name, argument):
+    name, model = settings.EMARSYS_EVENTS[event_name][argument]
+    return EventParam(
+        argument=argument,
+        name=name,
+        model=model
+    )
+
+
+def get_event_id(name):
+    try:
+        return Event.objects.get(name=name).emarsys_id
+    except Event.DoesNotExist:
+        pass
+
+    log.info("Emarsys-ID for event '{name}' unknown, "
+             "syncing...".format(name=name))
+
+    sync_events()
+
+    try:
+        return Event.objects.get(name=name).emarsys_id
+    except Event.DoesNotExist:
+        pass
+
+    return None
+
+
+def sync_events():
+    """
+    Sync locally available events, identified by their name, with emarsys.
+
+    In particular:
+        * Add events that don't exist locally.
+        * Update changed IDs.
+        * Delete Event objects if their name isn't reported by Emarsys
+          anymore. There are no foreign keys pointing to Event, so this
+          is safe.
+
+    Events are identified by their name. This allows adding and
+    implementing events in advance without knowing their emarsys id.
+    Syncing will then add a Event object with the emarsys ID so that they
+    can be triggered.
+
+    :return: num_new_events
+    """
+    num_new_events = 0
+    num_updated_ids = 0
+    num_deleted_ids = 0
+    unknown_event_names = []
+    unsynced_event_names = []
+
+    try:
+        emarsys_events = api.get_events()
+    except EmarsysError as e:
+        log.error('Emarsys error: {}\n'.format(e))
+        return num_new_events, num_updated_ids, num_deleted_ids, \
+            unknown_event_names, unsynced_event_names
+
+    for local_event in Event.objects.all():
+        emarsys_event_id = emarsys_events.get(local_event.name)
+        if emarsys_event_id:
+            if local_event.emarsys_id != emarsys_event_id:
+                num_updated_ids += 1
+
+            local_event.emarsys_id = emarsys_event_id
+            del emarsys_events[local_event.name]
+            local_event.save()
+        else:
+            local_event.delete()
+            num_deleted_ids += 1
+
+    for emarsys_event_name, emarsys_event_id in emarsys_events.items():
+        Event.objects.create(emarsys_id=emarsys_event_id,
+                             name=emarsys_event_name)
+        num_new_events += 1
+
+    all_synced_event_names = set(Event.objects.all()
+                                 .values_list('name', flat=True))
+    known_event_names = set(settings.EMARSYS_EVENTS.keys())
+    unknown_event_names = list(all_synced_event_names - known_event_names)
+    unsynced_event_names = list(known_event_names - all_synced_event_names)
+
+    if unknown_event_names:
+        log.warning("these Emarsys event names are not "
+                    "configured in settings.EMARSYS_EVENTS: {}"
+                    .format(', '.join(
+                        '"{}"'.format(x.replace('"', '\\"'))
+                            for x in unknown_event_names)))
```

Comment:
Ok. In that case `sync_events` could and probably should ignore all events it isn't configured for, so they never make it into the DB.